### PR TITLE
[SPARK-37895][SQL] Filter push down column with quoted columns

### PR DESCRIPTION
**Pull request by planga82**
### What changes were proposed in this pull request?

The problem happens when you have a column that is quoted because it has special characters.

```
select view1.`Имя1` , view1.`Имя2`, view2.`Имя3` 
from view1 left join  view2 
on view1.`Имя2`=view2.`Имя4`
```

It result in the following JDBC push down

```
SELECT "Имя3","Имя4" FROM "public"."tab2"  WHERE ("`Имя4`" IS NOT NULL) 
```

This PR fix this and in the push down query only remains the dialect quote

```
SELECT "Имя3","Имя4" FROM "public"."tab2"  WHERE ("Имя4" IS NOT NULL) 
```

### Why are the changes needed?

Fix this bug in the JDBC push down

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit testing

### CODE REVIEW

**Comment by planga82** [comment]
CC @cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalogSuite.scala [hunk 1]
```diff
@@ -425,4 +425,20 @@ class JDBCTableCatalogSuite extends QueryTest with SharedSparkSession {
       assert(m.contains("\"TABLEENGINENAME\" not found"))
     }
   }
+
+  test("JDBC push down with delimited special identifiers") {
```
Comment (huaxingao) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalogSuite.scala:429]: We actually have a `JDBCV2Suite`. It's probably better to put this test there.

Comment (planga82) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalogSuite.scala:429]: Test moved, thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala [hunk 2]
```diff
@@ -97,7 +97,15 @@ object JDBCRDD extends Logging {
    * Returns None for an unhandled filter.
    */
   def compileFilter(f: Filter, dialect: JdbcDialect): Option[String] = {
-    def quote(colName: String): String = dialect.quoteIdentifier(colName)
+    def isEnclosedInBackticks(colName: String): Boolean =
+      colName.startsWith("`") && colName.endsWith("`")
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala:101]: I checked `V2ScanRelationPushDown`. For v2 sources, Spark always quote the column name if it contains special chars for filter pushdown. So I think here we can just invoke the sql parser to parse it
```
val nameParts = SparkSession.active.sessionState.sqlParser.parseMultipartIdentifier(colName)
assert(nameParts.length == 1) // or throw a user-facing exception if nested column can reach here
dialect.quoteIdentifier(nameParts.head)
```

Comment (planga82) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCRDD.scala]: Great! I've changed it

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**